src/riak_kv_token_session.erl

```diff
@@ -0,0 +1,467 @@
+-define(TOKEN_SESSION_TIMEOUT, 30000).
+-define(TOKEN_RETRY_COUNT, 12).
+
+-record(state,
+        {
+            token_id :: token_id(),
+            token_timeout :: pos_integer(),
+            client :: riak_client:riak_client()|undefined,
+            session_id :: session_id()|undefined,
+            start_time = os:system_time(microsecond) :: pos_integer()
+        }
+    ).
+
+-type token_id() :: riak_kv_token_manager:token_id().
+-type verify_list() :: riak_kv_token_manager:verify_list().
+-type session_ref() :: binary().
+-type session_id() :: non_neg_integer().
+-type timeout_ms() :: pos_integer().
+-type erpc_error() :: {erpc, term()}.
+
+-export_type([session_ref/0]).
+
+
+%%%============================================================================
+%%% External API
+%%%============================================================================
+
+-spec session_request_retry(
+    token_id()) -> {true, session_ref()}|{false, none}|erpc_error().
+session_request_retry(TokenID) ->
+    NVal =
+        application:get_env(
+            riak_kv, stronger_conditional_nval, ?CONDITIONAL_NVAL
+        ),
+    session_request_retry(
+        TokenID,
+        NVal,
+        ?TOKEN_REQUEST_TIMEOUT,
+        ?TOKEN_SESSION_TIMEOUT,
+        ?TOKEN_RETRY_COUNT
+    ).
+
+-spec session_request_retry(
+    token_id(), 3|5, timeout_ms(), timeout_ms(), pos_integer())
+        -> {true, session_ref()}|{false, none}|erpc_error().
+session_request_retry(TokenID, NVal, RequestTimeout, TokenTimeout, Retry) ->
+    session_request_retry(
+        TokenID, NVal, RequestTimeout, TokenTimeout, Retry, 0
+    ).
+
+session_request_retry(_TokenID, _NVal, 0, _TokenTO, _Retry, _Attempts) ->
+    {false, none};
+session_request_retry(TokenID, NVal, RequestTO, TokenTO, Retry, Retry) ->
+    session_request(TokenID, NVal, RequestTO, TokenTO);
+session_request_retry(TokenID, NVal, RequestTO, TokenTO, Retry, Attempts) ->
+    RT0 = os:system_time(millisecond),
+    case session_request(TokenID, NVal, RequestTO, TokenTO) of
+        {true, SessionRef} ->
+            {true, SessionRef};
+        _Error ->
+            timer:sleep(
+                rand:uniform(max(1, RequestTO div (Retry - Attempts)))
+            ),
+            session_request_retry(
+                TokenID,
+                NVal,
+                max(0, RequestTO + (RT0 - os:system_time(millisecond))),
+                TokenTO,
+                Retry,
+                Attempts + 1
+            )
+    end.
+
+-spec session_request(token_id(), 1|3|5|7, timeout_ms(), timeout_ms())
+        -> {true, session_ref()}|{false, none}|erpc_error().
+session_request(TokenID, NVal, RequestTimeout, TokenTimeout)
+        when NVal ==1; NVal == 3; NVal == 5; NVal == 7 ->
+    DocIdx = chash_key(TokenID),
+    PrimPartitions = riak_core_apl:get_primary_apl(DocIdx, NVal, riak_kv),
+    TargetNodeCount =
+        case NVal of
+            1 -> 1;
+            TNC when TNC > 1 -> TNC -2
```

Comment:
I wonder about this hard coded 2.
Compared to the documentation in `token_manager` 
```
%% token.  The next two nodes in the preflist (that are UP), are responsible
%% for verifying that the grant can be made (i.e. to confirm that they have not
%% granted a token while the head node was recently unavailable). 
```
Isn't the NVal used in `token_session` actually Target Nval and should then the TargetNodeCount not be `Nval-1` for the actual NVal, not the TargetNVal?

Reply:
Changed as part of renaming the configuration options